MNT Add validation of scalar parameters in Ridge

ArturoAmorQ opened the pull request:
<!--
Thanks for contributing a pull request! Please ensure you have taken a look at
the contribution guidelines: https://github.com/scikit-learn/scikit-learn/blob/main/CONTRIBUTING.md
-->

#### Reference Issues/PRs
Solves a part of the Issue  #20724 as mentioned in [this comment](https://github.com/scikit-learn/scikit-learn/issues/20724#issuecomment-943344299)

#### What does this implement/fix? Explain your changes.

Use the helper function `check_scalar` from `sklearn.utils` to validate scalar parameters and making sure to get consistent error types and messages in `linear_model.Ridge`

#### Any other comments?


Happy to contribute to the scikit-learn project for the first time!

Commit:
Add validation on alpha

Commit:
Modify alpha inplace for validation

Commit:
Format code with pre-commit config

Commit:
Validate max_iter for Ridge

Also corrects validation for alpha

Commit:
Add tests for alpha and max_iter validation

Commit:
Validate tol for Ridge

Commit:
Add tests for tol validation

jjerphan commented on sklearn/linear_model/_ridge.py:
```diff
@@ -556,6 +558,15 @@ def _ridge_regression(
             # we implement sample_weight via a simple rescaling.
             X, y = _rescale_data(X, y, sample_weight)
 
+    if not isinstance(alpha, (np.ndarray, type(None), tuple)):
+        alpha = check_scalar(
+            alpha,
+            "alpha",
+            target_type=numbers.Real,
+            min_val=0.0,
+            include_boundaries="left",
+        )
```
Was this change was introduced for `RidgeCV`?

If so, I would reintroduce it in another PR and move the `check_scalar` call under the one for `self.max_iter` in `fit`.

If not, I would add this comment to explain its intend:
```suggestion
    # Some callers of this method might pass alpha as single
    # element array which already has been validated.
    if not isinstance(alpha, (np.ndarray, type(None), tuple)):
        alpha = check_scalar(
            alpha,
            "alpha",
            target_type=numbers.Real,
            min_val=0.0,
            include_boundaries="left",
        )
```

jjerphan commented on sklearn/linear_model/tests/test_ridge.py:
```diff
@@ -335,6 +335,47 @@ def test_ridge_individual_penalties():
+        ),
+    ],
+)
+def test_Ridge_params_validation(params, err_type, err_msg):
+    """Check the parameters validation in `Ridge`."""
```
`snake_case` must be use to name for function, hence no uppercase even if the name is constructed from the one of a class (see [this section of PEP8](https://www.python.org/dev/peps/pep-0008/#descriptive-naming-styles)). The name of the test (`test_X_CenterStackOp`) bellow is an exception.
 
Also and IIRC, an implicit convention is that back-quotes aren't use on single-lined doc-string.
 
```suggestion
def test_ridge_params_validation(params, err_type, err_msg):
    """Check the parameters validation in Ridge."""
```

jjerphan reviewed (commented):
Just a few comments; apart from those, this looks good to me.

ArturoAmorQ committed:
Update sklearn/linear_model/_ridge.py

This change is indeed independent of `RidgeCV`. This commit adds an explanation on the need for a conditional statement during validation.

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

ArturoAmorQ committed:
Update sklearn/linear_model/tests/test_ridge.py

Tweak to adopt conventions

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

ogrisel commented on sklearn/linear_model/_ridge.py:
```diff
@@ -556,6 +558,17 @@ def _ridge_regression(
 
+    # Some callers of this method might pass alpha as single
+    # element array which already has been validated.
+    if not isinstance(alpha, (np.ndarray, type(None), tuple)):
```
nitpick, I find the following more direct and easy to understand:

```suggestion
    if alpha is not None and not isinstance(alpha, (np.ndarray, tuple)):
```

ogrisel commented on sklearn/linear_model/tests/test_ridge.py:
```diff
@@ -335,6 +335,47 @@ def test_ridge_individual_penalties():
+        (
+            {"tol": 1},
+            TypeError,
+            "tol must be an instance of <class 'float'>, not <class 'int'>",
```
I think we could accept any `numbers.Real` to automatically treat int as floats.

ArturoAmorQ replied:
Done, thanks!

ogrisel reviewed (approved):
LGTM once the following comments are addressed.

I think this deserves a changelog entry in `doc/whats_new/v1.1.rst` in the `linear_model` section stating that Ridge and ridge_regression now raise consistent error message when passed invalid values for alpha, max_iter and tol.

glemaitre reviewed (approved):
It looks good but I would prefer to merge the PR moving the validation first.

ArturoAmorQ committed:
Improve readability

Co-authored-by: Olivier Grisel <olivier.grisel@ensta.org>

Commit:
Change `tol` from <class 'float'> to <class 'numbers.Real'>

Commit:
Edit tests to accept `tol` in <class 'float'>

Commit:
Add changelog entry in `doc/whats_new/v1.1.rst`

Commit:
Merge branch 'main' into validations_for_ridge

jjerphan commented on doc/whats_new/v1.1.rst:
```diff
@@ -141,6 +141,12 @@ Changelog
   multilabel classification.
   :pr:`19689` by :user:`Guillaume Lemaitre <glemaitre>`.
 
+- |Enhancement| :class:`linear_model.Ridge` and :class:`linear_model.RidgeClassifier`
+  now raise consistent error message when passed invalid values for `alpha`,
+  `max_iter` and `tol`.
+  :pr:`21341` by :user:`Arturo Amor <ArturoAmorQ>`
+  and :user:`Julien Jerphanion <jjerphan>`.
```
_You_ did this. :wink: 
```suggestion
- |Enhancement| :class:`linear_model.Ridge` and :class:`linear_model.RidgeClassifier`
  now raise consistent error message when passed invalid values for `alpha`,
  `max_iter` and `tol`.
  :pr:`21341` by :user:`Arturo Amor <ArturoAmorQ>`.
```

jjerphan reviewed (approved):
LGTM once the conflicts with `main` are solved. Thank you, @ArturoAmorQ!

glemaitre commented:
@ArturoAmorQ Could you solve the conflict? We will be able to merge this one.

ArturoAmorQ committed:
Update doc/whats_new/v1.1.rst

Co-authored-by: Julien Jerphanion <git@jjerphan.xyz>

Commit:
Merge branch 'main' into validations_for_ridge

Commit:
tweak
